File: scripts/fork_purger.py
"""
To test the script locally, perform the following steps:

* Create a Python 3.11+ virtual environment.
* Install the dependencies.
* Export the following environment variables:
    * GITHUB_USERNAME
    * GITHUB_API_TOKEN
    * TIMEOUT
* Run the script:
    ```
    python -m scripts.purge_forks
    ```
"""

from __future__ import annotations
import httpx
import os
import asyncio
import logging
from contextlib import AsyncExitStack
import datetime


# Get the GitHub API token from the environment.
GITHUB_API_TOKEN = os.environ.get("GITHUB_API_TOKEN")
GITHUB_USERNAME = os.environ.get("GITHUB_USERNAME")

# The script will raise an exception if the timeout is reached.
TIMEOUT = float(timeout) if (timeout := os.environ.get("TIMEOUT")) else None

# Only delete forks that haven't been updated in the last 60 days.
OLDER_THAN = int(older_than) if (older_than := os.environ.get("OLDER_THAN")) else 60

# Raise runtime error if any of the environment variables are missing.
if all(
    None is env_var
    for env_var in (GITHUB_API_TOKEN, GITHUB_USERNAME, TIMEOUT, OLDER_THAN)
):
    raise RuntimeError(
        "Missing environment variables. Please set 'GITHUB_API_TOKEN', 'GITHUB_USERNAME', 'TIMEOUT', and 'OLDER_THAN'.",
    )

# Configure basic logger.
logging.basicConfig(level=logging.INFO)


async def get_forked_repos(
    http_client: httpx.AsyncClient,
    older_than: int = OLDER_THAN,  # 60 days.
) -> list[str]:
    """Get a list of all forked repositories.

    Args:
        http_client (httpx.AsyncClient): HTTP client.
        older_than (int): Only collect repos that are forked before 'older_than' days.
    """

    logging.info("Getting the list of forked repositories...")
    # Get the list of repositories.
    response = await http_client.head(
        f"https://api.github.com/users/{GITHUB_USERNAME}/repos",
        headers={
            "Accept": "application/vnd.github+json",
            "Authorization": f"Bearer {GITHUB_API_TOKEN}",
        },
    )
    # Raise an exception if the request failed.
    response.raise_for_status()

    # Get the number of paginated urls.
    num_pages = int(
        response.headers["link"]
        .split(",")[-1]
        .split(";")[0]
        .split("=")[-1]
        .replace(">", "")
    )

    forked_repo_urls = []
    for current_page in range(1, num_pages + 1):
        # Get the list of repositories.
        response = await http_client.get(
            f"https://api.github.com/users/{GITHUB_USERNAME}/repos?page={current_page}",
            headers={
                "Accept": "application/vnd.github+json",
                "Authorization": f"Bearer {GITHUB_API_TOKEN}",
            },
        )
        # Raise an exception if the request failed.
        response.raise_for_status()

        # Get current UTC datetime.
        current_dt = datetime.datetime.now(datetime.UTC)

        # 60 days ago.
        older_than_dt = current_dt - datetime.timedelta(days=older_than)

        for repo in response.json():
            if not repo["fork"]:
                continue

            # Get the datetime when the repo was forked.
            last_updated_dt = datetime.datetime.strptime(
                repo["updated_at"], "%Y-%m-%dT%H:%M:%SZ"
            )

            # Add tzinfo to the datetime.
            last_updated_dt = last_updated_dt.replace(tzinfo=datetime.UTC)

            # If the forked repo was last updated before 60 days, add it to the list.
            if last_updated_dt < older_than_dt:
                forked_repo_urls.append(repo["url"])

    # Return the list of forked repositories.
    return forked_repo_urls


async def delete_forked_repo(http_client: httpx.AsyncClient, repo_url: str) -> None:
    """Delete a forked repository."""

    # Delete the repository.
    response = await http_client.delete(
        repo_url,
        headers={
            "Accept": "application/vnd.github+json",
            "Authorization": f"Bearer {GITHUB_API_TOKEN}",
        },
    )
    # Raise an exception if the request failed.
    response.raise_for_status()
    logging.info(f"Deleted '{repo_url}'.")


async def main() -> None:
    """Main function."""

    # Create an HTTP client.
    http_client = httpx.AsyncClient()

    # Get the list of forked repositories.
    forked_repo_urls = await get_forked_repos(http_client)

    # If there are no forked repositories that were updated more than 60 days ago, exit.
    if not forked_repo_urls:
        logging.info(
            "No forked repositories found that were updated more than %s days ago.",
            OLDER_THAN,
        )
        return

    async with AsyncExitStack() as stack:
        await stack.enter_async_context(asyncio.timeout(TIMEOUT))
        task_group = await stack.enter_async_context(asyncio.TaskGroup())

        logging.info("Deleting forked repositories...")
        for repo_url in forked_repo_urls:
            task_group.create_task(delete_forked_repo(http_client, repo_url))


if __name__ == "__main__":
    asyncio.run(main())

File: scripts/fork_purger.go
/*
Script to delete all forked repos older than 60 days. To test it locally:

- Ensure you have go 1.21+ installed
- Install the dependencies with `go get ./scripts/...`
- The following env vars are set:
	- GH_TOKEN: a personal access token with repo delete access
	- GH_USERNAME: the username of the account to delete forked repos from
- Run `go run scripts/fork_purger.go`
*/

package main

import (
	"context"
	"log"
	"os"
	"time"

	"github.com/google/go-github/v53/github"
	"golang.org/x/oauth2"
)

const (
	timeout   = 1 * time.Second
	olderThan = 24 * 60 * time.Hour
	perPage   = 100
)

var (
	githubAPIToken = os.Getenv("GH_TOKEN")
	githubUsername = os.Getenv("GH_USERNAME")
)

func init() {
	// Validate required env vars
	if githubAPIToken == "" || githubUsername == "" {
		log.Fatal("GH_TOKEN and GH_USERNAME must be set")
	}
}

func main() {
	// Create context with timeout
	ctx, cancel := context.WithTimeout(context.Background(), timeout)
	defer cancel()

	// Authenticate client
	client := auth(ctx, githubAPIToken)

	// Get forked repos older than threshold
	repos, err := getForkedRepos(ctx, client, githubUsername, olderThan)
	if err != nil {
		log.Fatalf("could not get forked repos: %v", err)
	}

	// Log if no repos to delete
	if len(repos) == 0 {
		log.Println("no forked repos to delete")
		return
	}

	// Delete identified repos
	if err := deleteForkedRepos(ctx, client, repos); err != nil {
		log.Fatalf("could not delete repos: %v", err)
	}

	// Log results
	log.Printf(
		"deleted %d forked repos that are older than %d\n",
		len(repos), olderThan/(24*time.Hour),
	)
}

// auth creates an authenticated client
func auth(ctx context.Context, token string) *github.Client {
	ts := oauth2.StaticTokenSource(
		&oauth2.Token{AccessToken: token},
	)
	tc := oauth2.NewClient(ctx, ts)
	return github.NewClient(tc)
}

// getForkedRepos returns all forked repos older than the given threshold
func getForkedRepos(
	ctx context.Context,
	client *github.Client,
	username string,
	olderThan time.Duration,
) ([]*github.Repository, error) {

	opt := &github.RepositoryListOptions{
		ListOptions: github.ListOptions{
			PerPage: perPage,
		},
	}

	var repos []*github.Repository

	for {
		results, resp, err := client.Repositories.List(ctx, username, opt)

		if err != nil {
			return nil, err
		}

		for _, repo := range results {
			if *repo.Fork && time.Since(repo.CreatedAt.Time) > olderThan {
				repos = append(repos, repo)
			}
		}

		if resp.NextPage == 0 {
			break
		}
		opt.Page = resp.NextPage
	}
	return repos, nil
}

// deleteForkedRepos deletes the given repos
func deleteForkedRepos(
	ctx context.Context, client *github.Client, repos []*github.Repository) error {

	for _, repo := range repos {
		_, err := client.Repositories.Delete(ctx, *repo.Owner.Login, *repo.Name)
		if err != nil {
			return err
		}
	}
	return nil
}
